core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -696,4 +701,102 @@ public ThreadContext getThreadContext() {
+
+        /**
+         * Check if the execution has been cancelled
+         * @return true if cancelled
+         */
+        boolean isCancelled();
+    }
+
+    /**
+     * This class encapsulates the scheduling of a {@link Runnable} that needs to be repeated on a interval. For example, checking a value
+     * for cleanup every second could be done by passing in a Runnable that can perform the check and the specified interval between
+     * executions of this runnable. <em>NOTE:</em> the runnable is only rescheduled to run again after completion of the runnable.
+     *
+     * For this class, <i>completion</i> means that the call to {@link Runnable#run()} returned or an exception was thrown and caught. In
+     * case of an exception, this class will log the exception and reschedule the runnable for its next execution. This differs from the
+     * {@link ScheduledThreadPoolExecutor#scheduleWithFixedDelay(Runnable, long, long, TimeUnit)} semantics as an exception there would
+     * terminate the rescheduling of the runnable.
+     */
+    static class ReschedulingRunnable extends AbstractRunnable implements Cancellable {
+
+        private final Runnable runnable;
+        private final TimeValue interval;
+        private final String executor;
+        private final ThreadPool threadPool;
+
+        private volatile boolean run = true;
+
+        /**
+         * Creates a new rescheduling runnable and schedules the first execution to occur after the interval specified
+         *
+         * @param runnable the {@link Runnable} that should be executed periodically
+         * @param interval the time interval between executions
+         * @param executor the executor where this runnable should be scheduled to run
+         * @param threadPool the {@link ThreadPool} instance to use for scheduling
+         */
+        ReschedulingRunnable(Runnable runnable, TimeValue interval, String executor, ThreadPool threadPool) {
+            this.runnable = runnable;
+            this.interval = interval;
+            this.executor = executor;
+            this.threadPool = threadPool;
+            threadPool.schedule(interval, executor, this);
+        }
+
+        @Override
+        public void cancel() {
+            run = false;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return run == false;
+        }
+
+        @Override
+        public void doRun() {
+            // always check run here since this may have been cancelled since the last execution and we do not want to run
+            if (run) {
+                runnable.run();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            threadPool.logger.warn("failed to run scheduled task [{}] on thread pool [{}]", t, runnable.toString(), executor);
+        }
+
+        @Override
+        public void onRejection(Throwable t) {
+            if (threadPool.logger.isDebugEnabled()) {
```

Comment:
I think we should cancel here, no?  currently, the onAfter will be called resulting in an endless loop.

also, if we're shutting down, this might result in needless warnings, so maybe check if we're canceled and if so only log  a debug log instead of warn?